File: android/src/main/kotlin/com/lexisnexis/risk/flyreel_sdk_flutter/FlyreelSdkFlutterPlugin.kt
package com.lexisnexis.risk.flyreel_sdk_flutter

import android.app.Application
import android.content.Context
import android.net.Uri
import com.lexisnexis.risk.flyreel.Flyreel
import com.lexisnexis.risk.flyreel.FlyreelAnalyticEvent
import com.lexisnexis.risk.flyreel.FlyreelAnalyticUser
import com.lexisnexis.risk.flyreel.FlyreelConfiguration
import com.lexisnexis.risk.flyreel.FlyreelCoordination
import com.lexisnexis.risk.flyreel.FlyreelDeviceData
import com.lexisnexis.risk.flyreel.FlyreelEnvironment
import com.lexisnexis.risk.flyreel.FlyreelMessageDetails
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlyreelSdkFlutterPlugin */
class FlyreelSdkFlutterPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null
    private var currentActivity: android.app.Activity? = null

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val arguments = call.arguments as Map<*, *>
                val organizationId = arguments["organizationId"] as String
                val settingsVersion = arguments["settingsVersion"] as Int
                val environment = arguments["environment"] as String

                Flyreel.initialize(
                    context as Application, FlyreelConfiguration(
                        organizationId = organizationId,
                        settingsVersion = settingsVersion,
                        environment = mapEnvironment(environment)
                    )
                )
                result.success(null)
            }

            "open" -> {
                val arguments = call.arguments as Map<*, *>
                val deeplinkUrl = arguments["deeplinkUrl"] as? String
                val shouldSkipLoginPage = arguments["shouldSkipLoginPage"] as Boolean
                (currentActivity as? Context)?.let {
                    Flyreel.openFlyreel(
                        context = it,
                        deeplinkUri = deeplinkUrl?.let { url -> Uri.parse(url) },
                        shouldSkipLoginPage = shouldSkipLoginPage
                    )
                    result.success(null)
                } ?: result.notImplemented()
            }

            "openWithCredentials" -> {
                val arguments = call.arguments as Map<*, *>
                val zipCode = arguments["zipCode"] as String
                val accessCode = arguments["accessCode"] as String
                val shouldSkipLoginPage = arguments["shouldSkipLoginPage"] as Boolean
                (currentActivity as? Context)?.let {
                    Flyreel.openFlyreel(
                        context = it,
                        zipCode = zipCode,
                        accessCode = accessCode,
                        shouldSkipLoginPage = shouldSkipLoginPage
                    )
                    result.success(null)
                } ?: result.notImplemented()
            }

            "checkStatus" -> {
                val arguments = call.arguments as Map<*, *>
                val zipCode = arguments["zipCode"] as String
                val accessCode = arguments["accessCode"] as String

                Flyreel.fetchFlyreelStatus(
                    zipCode = zipCode,
                    accessCode = accessCode,
                    onSuccess = { status ->
                        result.success(
                            mapOf(
                                "status" to status.status,
                                "expiration" to status.expiration
                            )
                        )
                    },
                    onError = { error ->
                        result.error(error.code.toString(), error.message, null)
                    }
                )
            }

            "enableLogs" -> {
                Flyreel.enableLogs()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun mapEnvironment(environment: String) =
        when (environment) {
            "production" -> FlyreelEnvironment.Production
            "sandbox" -> FlyreelEnvironment.Sandbox
            "staging" -> FlyreelEnvironment.Staging
            else -> throw Exception("Wrong Flyreel environment")
        }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flyreel_sdk_flutter")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flyreel_sdk_stream")
            .also {
                it.setStreamHandler(this)
            }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
        eventSink = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        currentActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
    }

    override fun onDetachedFromActivity() {
        currentActivity = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
        Flyreel.observeAnalyticEvents { event ->
            eventSink?.success(event.toMap())
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        eventChannel = null
    }
}

internal fun FlyreelAnalyticEvent.toMap(): Map<String, Any?> {
        return mapOf(
            "user" to user.toMap(),
            "name" to name,
            "timestamp" to timestamp,
            "activeTime" to activeTime,
            "coordination" to coordination?.toMap(),
            "deviceData" to deviceData?.toMap(),
            "messageDetails" to messageDetails?.toMap()
        )
    }

internal fun FlyreelAnalyticUser.toMap(): Map<String, Any?> {
    return mapOf(
        "id" to id,
        "name" to name,
        "email" to email,
        "botId" to botId,
        "botName" to botName,
        "organizationId" to organizationId,
        "status" to status,
        "loginType" to loginType.value
    )
}

internal fun FlyreelDeviceData.toMap(): Map<String, Any?> {
    return mapOf(
        "phoneManufacturer" to phoneManufacturer,
        "phoneModel" to phoneModel,
        "appVersion" to appVersion,
        "appName" to appName
    )
}

internal fun FlyreelCoordination.toMap(): Map<String, Any?> {
    return mapOf(
        "lat" to lat,
        "lng" to lng
    )
}

internal fun FlyreelMessageDetails.toMap(): Map<String, Any?> {
    return mapOf(
        "message" to message,
        "messageType" to messageType,
        "moduleKey" to moduleKey,
        "messageKey" to messageKey
    )
}

File: ios/Classes/FlyreelSdkFlutterPlugin.swift
import Flutter
import UIKit
@_spi(FlyreelInternal) import Flyreel

public class FlyreelSdkFlutterPlugin: NSObject, FlutterPlugin {
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flyreel_sdk_flutter", binaryMessenger: registrar.messenger())
        let instance = FlyreelSdkFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        let eventChannel = FlutterEventChannel(name: "flyreel_sdk_stream", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(AnalyticsHandler())
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            let arguments = call.arguments as! [String: Any]
            let organizationId = arguments["organizationId"] as! String
            let settingsVersion = arguments["settingsVersion"] as! Int
            let environmentString = arguments["environment"] as! String
            
            let configuration = FlyreelConfiguration(
                settingsVersion: String(settingsVersion),
                organizationId: organizationId,
                environment: mapEnvironment(environment: environmentString)
            )
            
            FlyreelSDK.shared.set(configuration: configuration)
            result(nil)
        case "open":
            let rootView = UIApplication.shared.delegate!.window!!.rootViewController!
            let arguments = call.arguments as! [String: Any]
            let deeplink = arguments["deeplinkUrl"] as? String
            let shouldSkipLoginPage = arguments["shouldSkipLoginPage"] as! Bool
            
            if let deeplinkURL = URL(string: deeplink ?? "") {
                FlyreelSDK.shared.presentFlyreel(on: rootView, deepLinkURL: deeplinkURL, shouldSkipLoginPage: shouldSkipLoginPage)
            } else {
                FlyreelSDK.shared.presentFlyreel(on: rootView)
            }
            result(nil)
        case "openWithCredentials":
            let rootView = UIApplication.shared.delegate!.window!!.rootViewController!
            let arguments = call.arguments as! [String: Any]
            let zipCode = arguments["zipCode"] as! String
            let accessCode = arguments["accessCode"] as! String
            let shouldSkipLoginPage = arguments["shouldSkipLoginPage"] as! Bool
            FlyreelSDK.shared.presentFlyreel(on: rootView, zipCode: zipCode, accessCode: accessCode, shouldSkipLoginPage: shouldSkipLoginPage)
            
            result(nil)
        case "enableLogs":
            FlyreelSDK.shared.enableLogs()
            result(nil)
        case "checkStatus":
            let arguments = call.arguments as! [String: Any]
            let zipCode = arguments["zipCode"] as! String
            let accessCode = arguments["accessCode"] as! String
            FlyreelSDK.shared.fetchFlyreelStatus(zipCode: zipCode, accessCode: accessCode) { statusResult in
                switch statusResult {
                case .success(let flyreelStatus):
                    result(["status": flyreelStatus.status, "expiration": flyreelStatus.expiration])
                case .failure(let error):
                    if case FlyreelError.apiError(let message, let code) = error {
                        result(FlutterError(code: String(code), message: message, details: nil))
                    } else {
                        result(FlutterError(code: "500", message: "Unexpected error", details: nil))
                    }
                }
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
 
    func mapEnvironment(environment: String) -> FlyreelEnvironment {
        switch environment {
        case "production":
            return .production
        case "sandbox":
            return .sandbox
        case "staging":
            return .staging
        default:
            return .production
        }
    }
}

class AnalyticsHandler: NSObject, FlutterStreamHandler {

       private var eventSink: FlutterEventSink?
       
       func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
           self.eventSink = eventSink
   
           FlyreelSDK.shared.observeAnalyticEvents { event in
               eventSink(event.toFlutterMap())
           }
           
       
           return nil
       }
       
       func onCancel(withArguments arguments: Any?) -> FlutterError? {
           eventSink = nil
           return nil
       }
}

extension FlyreelAnalyticEvent {
    func toFlutterMap() -> [String: Any?] {
        return [
            "user": user?.toFlutterMap(),
            "name": eventName,
            "timestamp": ISO8601DateFormatter().string(from: timestamp),
            "activeTime": activeTimeFromTrigger(),
            "coordination": coordinationFromTrigger()?.toFlutterMap(),
            "deviceData": deviceDataFromTrigger()?.toFlutterMap(),
            "messageDetails": messageDetailsFromTrigger()?.toFlutterMap()
        ]
    }

    private func activeTimeFromTrigger() -> TimeInterval? {
        switch trigger {
        case .questionAsked(let activeTime, _),
             .questionAnswered(let activeTime, _, _),
             .flyreelCompleted(let activeTime):
            return activeTime
        case .userLoggedIn:
            return nil
        }
    }

    private func coordinationFromTrigger() -> FlyreelCoordinate? {
        switch trigger {
        case .questionAnswered(_, let location, _),
             .userLoggedIn(_, _, let location, _):
            return location
        default:
            return nil
        }
    }

    private func deviceDataFromTrigger() -> FlyreelDeviceData? {
        switch trigger {
        case .userLoggedIn(_, let deviceData, _, _):
            return deviceData
        default:
            return nil
        }
    }

    private func messageDetailsFromTrigger() -> AnalyticMessageDetails? {
        switch trigger {
        case .questionAsked(_, let message),
             .questionAnswered(_, _, let message):
            return message
        default:
            return nil
        }
    }
}

extension FlyreelDeviceData {
    func toFlutterMap() -> [String: String?] {
        return [
            "phoneModel": phoneModel,
            "appVersion": appVersion,
            "appName": appName
        ]
    }
}

extension FlyreelCoordinate {
    func toFlutterMap() -> [String: Any] {
        return [
            "lat": lat,
            "lng": lng
        ]
    }
}

extension AnalyticMessageDetails {
    func toFlutterMap() -> [String: Any] {
        return [
            "message": message,
            "messageType": messageType,
            "moduleKey": moduleKey,
            "messageKey": messageKey,
            "phase": phase
        ]
    }
}

extension FlyreelAnalyticUser {
    func toFlutterMap() -> [String: Any?] {
        return [
            "id": flyreelID,
            "name": name,
            "email": email,
            "botId": botId,
            "botName": botName,
            "organizationId": organizationId,
            "status": status
        ]
    }
}
